sretoolbox/utils/multiprocess.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2021 Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Multiprocessing abstractions.
+"""
+
+from multiprocessing import Pool
+
+from sretoolbox.utils.exception import SystemExitWrapper
+
+
+def run(func, iterable, process_pool_size,
+        return_exceptions=False, **kwargs):
+    """
+    run_processes executes a function for each item in the input iterable.
+    execution will be done in a processpool according to the input
+    process_pool_size.  kwargs are passed to the input function
+    (optional). If return_exceptions is true, any exceptions that may
+    have happened in each thread are returned in the return value,
+    allowing the caller to get as much work done as possible.
+
+    SystemExit exceptions are treated the same way as regular exceptions.
+    """
+    if return_exceptions:
+        tracer = _catching_traceback
+    else:
+        tracer = _full_traceback
+
+    task_list = [(func, [i], kwargs) for i in iterable]
+    with Pool(process_pool_size) as pool:
+        try:
+            return pool.map(tracer, task_list)
+        except SystemExitWrapper as details:
+            # a SystemExitWrapper is just a wrapper around a SystemExit
+            # so we can catch it here reliably and propagate the actual
+            # SystemExit as is
+            raise details.origional_sys_exit_exception
+
+
+def _catching_traceback(spec):
+    try:
+        func, args, kwargs = spec
+        return func(*args, **kwargs)
+    # pylint: disable=broad-except
+    except Exception as details:
+        return details
+    except SystemExit as details:
```

Comment:
Should we merge this into a `except BaseException`? That would also catch these extra exceptions:

- BaseExceptionGroup
- GeneratorExit
- KeyboardInterrupt

But that seems fine, I guess. Wdyt?

Reply:
i was thinking about this but was undecided. we definitely can and it is most likely the safe thing to do! ACK

Reply:
[182b878](https://github.com/app-sre/sretoolbox/pull/89/commits/182b8785ef99c7a3f9071e6585476ce684d558bd)